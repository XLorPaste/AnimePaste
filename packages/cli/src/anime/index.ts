export * from './anime';

export * from './search';

export * from './utils';
