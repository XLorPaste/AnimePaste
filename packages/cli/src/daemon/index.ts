import { Daemon } from './daemon';

export async function startDaemon(option: {
  once: boolean;
  interval: number;
  update: boolean;
}): Promise<void> {
  const daemon = new Daemon(option);

  await daemon.init();

  return new Promise((res) => {
    const stamp = setInterval(async () => {
      await daemon.update();

      if (option.once) {
        clearInterval(stamp);
        res();
      }
    }, option.interval * 60 * 1000);
  });
}
