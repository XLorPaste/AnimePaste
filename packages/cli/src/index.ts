export * from './anime';

export * from './client';

export * from './context';

export * from './daemon';

export * from './io';

export * from './video';
