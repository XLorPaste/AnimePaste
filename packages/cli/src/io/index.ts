export * from './types';

export * from './utils';

export * from './download';

export * from './store';

export * from './torrent';

export * from './video';
