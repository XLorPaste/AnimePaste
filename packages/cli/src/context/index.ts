import * as fs from 'fs-extra';
import * as path from 'node:path';
import { homedir } from 'node:os';

import { subMonths } from 'date-fns';
import { load, dump } from 'js-yaml';

import { Database } from '@animepaste/database';

import type { LocalVideoInfo, CliOption, Plan } from '../types';

import { Anime } from '../anime';

import { LogContext } from './log';

export interface GlobalConfig {
  plan: string | string[];

  server: {
    baseURL: string;

    token: string;
  };

  store: {};
}

const DefaultGlobalConfig: GlobalConfig = {
  plan: ['./plans/test.yaml'],
  server: {
    baseURL: 'https://anime.xlorpaste.cn/api/',
    token: ''
  },
  store: {
    local: {
      path: './anime'
    },
    ali: {
      accessKeyId: '',
      accessKeySecret: '',
      endpoint: ''
    }
  }
};

export class GlobalContex {
  static ConfigFileName = 'config.yaml';
  static AnimeDdName = 'anime.json';

  cliOption!: CliOption;

  readonly root: string;
  readonly anime: string;
  readonly config: string;
  readonly cacheRoot: string;
  readonly databaseFilepath: string;

  readonly storeLog: LogContext<LocalVideoInfo>;
  readonly database: Database;

  private _localRoot: string;
  private configCache: any;
  private animeCache: Map<string, Anime> = new Map();

  constructor() {
    this.root = path.join(homedir(), '.animepaste');
    this.cacheRoot = path.join(this.root, 'cache');
    this._localRoot = path.join(this.root, 'anime');
    this.databaseFilepath = path.join(this.root, 'magnet.db');

    this.anime = path.join(this.root, GlobalContex.AnimeDdName);
    this.config = path.join(this.root, GlobalContex.ConfigFileName);

    this.storeLog = new LogContext(this, 'store.json');
    this.database = new Database({ url: this.databaseFilepath });
  }

  async init(option: CliOption) {
    this.cliOption = option;

    await fs.ensureDir(this.root);
    await fs.ensureDir(path.join(this.root, 'anime'));
    await fs.ensureDir(path.join(this.root, 'cache'));
    await fs.ensureDir(path.join(this.root, 'plans'));

    if (!(await fs.pathExists(this.config))) {
      fs.writeFile(
        this.config,
        dump(DefaultGlobalConfig, { indent: 2 })
          .replace('server', '\nserver')
          .replace('store', '\nstore'),
        'utf-8'
      );
    }

    const local = await this.getStoreConfig<{ path: string }>('local');
    if (local?.path) {
      this._localRoot = path.resolve(this.root, local.path);
    }
    if (!fs.existsSync(this._localRoot)) {
      throw new Error(`Local stroage root "${this._localRoot}" does not exist`);
    }

    if (fs.existsSync(this.anime)) {
      const animes = JSON.parse(
        await fs.readFile(this.anime, 'utf-8')
      ) as Anime[];
      this.animeCache.clear();
      for (const anime of animes) {
        this.animeCache.set(anime.bgmId, Anime.copy(anime));
      }
    }

    await this.database.init();
    await this.loadConfig();
  }

  get localRoot() {
    return this._localRoot;
  }

  async makeLocalAnimeRoot(title: string) {
    const local = path.join(this.localRoot, title);
    await fs.ensureDir(local);
    return local;
  }

  async loadConfig<T = any>(): Promise<T> {
    const content = await fs.readFile(this.config, 'utf-8');
    this.configCache = load(content);
    return this.configCache;
  }

  async getPlans(): Promise<Plan[]> {
    const config = await this.loadConfig<GlobalConfig>();
    const planPath = Array.isArray(config.plan) ? config.plan : [config.plan];
    const planBody: Plan[] = [];
    for (const plan of planPath) {
      const planPath = path.join(this.root, plan);
      if (!fs.existsSync(planPath)) {
        throw new Error(`You should provide plan "${plan}"`);
      }
      const body = load(fs.readFileSync(planPath, 'utf-8')) as Plan;
      // Setup date (default: 6 months ago)
      if (!Boolean(body.date)) {
        body.date = subMonths(new Date(), 6);
      } else {
        body.date = new Date(body.date);
      }
      // Setup state (default: onair)
      if (!Boolean(body.state)) {
        body.state = 'onair';
      }
      planBody.push(body);
    }
    return planBody;
  }

  async getServerConfig(): Promise<GlobalConfig['server']> {
    const config = await this.loadConfig<GlobalConfig>();
    return config.server;
  }

  async getStoreConfig<T = any>(key: string): Promise<T> {
    return (await this.loadConfig()).store[key];
  }

  // -----------

  /**
   * Anime
   */
  async getAnime(bgmId: string): Promise<Anime | undefined> {
    return this.animeCache.get(bgmId);
  }

  async updateAnime(anime: Anime) {
    this.animeCache.set(anime.bgmId, anime);
    const content = [...this.animeCache.values()];
    await fs.writeFile(this.anime, JSON.stringify(content, null, 2), 'utf-8');
  }

  // -----------

  /**
   * Copy file from "src" to "root/dst/basename(src)"
   *
   * @param src
   * @param dst
   */
  async copy(src: string, dst: 'cache' | 'anime') {
    const filepath = path.join(this.root, dst, path.basename(src));
    await fs.copy(src, filepath);
    return filepath;
  }
}

export const context = new GlobalContex();
