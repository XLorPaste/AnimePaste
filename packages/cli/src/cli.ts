import path from 'node:path';
import { execSync } from 'node:child_process';
import { existsSync, readFileSync, remove } from 'fs-extra';

import Breadc from 'breadc';
import { debug as createDebug } from 'debug';
import { lightRed, red, link, green, dim, lightBlue } from 'kolorist';

import type { AnimeType } from './types';

import { context } from './context';
import { printVideoInfo } from './io';
import { IndexListener, printMagnets, padRight } from './logger';

const name = 'anime';

const debug = createDebug(name + ':cli');

const cli = Breadc(name, {
  version: getVersion(),
  description: 'Paste your favourite anime online.',
  logger: { debug }
}).option('-f, --force', 'Enable force mode and prefer not using cache');

cli
  .command('watch', 'Watch anime resources update')
  .option('-i, --interval [duration]', 'Damon interval in minutes', {
    construct(t) {
      return t ? +t : 10;
    }
  })
  .option('-o, --once', 'Just do an immediate update')
  .option('--update', 'Only update info')
  .action(async (option) => {
    const { startDaemon } = await import('./daemon');
    await startDaemon(option);
  });

cli
  .command('plan')
  .option('--type [type]', 'One of local or server')
  .action(async (option) => {
    const { Plan } = await import('./daemon');
    const plan = await Plan.create();
    const type = option.type ?? 'local';
    if (type === 'local') {
      plan.printOnair();
    } else if (type === 'server') {
      plan.printOnair();
    }
  });

cli
  .command('search [anime]', 'Search Bangumi resources')
  .option('--type [type]', {
    construct(t) {
      if (t && ['tv', 'web', 'movie', 'ova'].includes(t)) {
        return t as AnimeType;
      } else {
        return 'tv';
      }
    }
  })
  .option('--raw', 'Print raw magnets')
  .option('--index', 'Index magnet database')
  .option('-y, --year [year]')
  .option('-m, --month [month]')
  .option('-p, --plan', 'Output plan.yaml')
  .action(async (anime, option) => {
    const { userSearch } = await import('./anime');
    if (option.index) {
      await context.magnetStore.index({ listener: IndexListener });
    }
    await userSearch(anime, option);
  });

cli
  .command(
    'fetch <id> <title> [...keywords]',
    'Fetch resources using Bangumi ID'
  )
  .option('--raw', 'Print raw magnets')
  .option('--index', 'Index magnet database')
  .option('-p, --plan', 'Output plan.yaml')
  .action(async (id, title, anime, option) => {
    const { daemonSearch } = await import('./anime');
    if (option.index) {
      await context.magnetStore.index({ listener: IndexListener });
    }
    await daemonSearch(id, [title, ...anime], {
      ...option,
      title,
      type: 'tv' as 'tv'
    });
  });

cli
  .command('store list [name]', 'List all uploaded video info')
  .alias('store ls')
  .option('--one-line', 'Only show one line')
  .action(async (name, option) => {
    const { useStore } = await import('./io');
    const store = await useStore('ali')();

    const videos = await store.listLocalVideos();
    videos.sort((a, b) => a.title.localeCompare(b.title));

    const logs: string[] = [];
    const ids: string[] = [];
    for (const info of videos) {
      if (!name || info.title.indexOf(name) !== -1) {
        if (option['one-line']) {
          logs.push(info.videoId);
        } else {
          logs.push(`  ${info.title}`);
          ids.push(`(${link(info.videoId, info.playUrl[0])})`);
        }
      }
    }

    if (option['one-line']) {
      console.log(logs.join(' '));
    } else {
      const padded = padRight(logs);
      for (let i = 0; i < padded.length; i++) {
        console.log(`${padded[i]}  ${ids[i]}`);
      }
    }
  });

cli
  .command('store get <id>', 'View video info on OSS')
  .option('--file', 'Use videoId instead of filepath')
  .action(async (id, option) => {
    const { useStore, printVideoInfo } = await import('./io');
    const store = await useStore('ali')();

    const info = !option.file
      ? await store.fetchVideoInfo(id)
      : await store.searchLocalVideo(id);

    if (info) {
      printVideoInfo(info);
    } else {
      console.log(`  ${red(`✗ video "${id}" not found`)}`);
    }
  });

cli
  .command('store put <file>', 'Upload video to OSS')
  .action(async (filename) => {
    const { useStore, printVideoInfo } = await import('./io');
    const store = await useStore('ali')();

    try {
      const info = await store.upload(path.resolve(filename));
      if (info) {
        printVideoInfo(info);
      } else {
        throw new Error();
      }
    } catch (error) {
      console.log();
      console.log(`  ${red('✗ Fail')} uploading ${filename}`);
    }
  });

cli
  .command('store remove [...ids]', 'Remove video info on OSS')
  .alias('store rm')
  .option('--file', 'Use filepath instead of videoId')
  .option('--rm-local', 'Remove local files')
  .action(async (ids, option) => {
    const { useStore } = await import('./io');
    const store = await useStore('ali')();

    for (const id of ids) {
      const info = !option.file
        ? await store.fetchVideoInfo(id)
        : await store.searchLocalVideo(id);

      if (option['rm-local'] && info?.source.directory) {
        const filepath = context.decodePath(info?.source.directory);
        await remove(filepath);
      }

      console.log();
      if (info) {
        printVideoInfo(info);
        await store.deleteVideo(info.videoId);
        console.log(`  ${green(`✓ Delete    ${info.videoId} Ok`)}`);
      } else {
        console.log(`  ${red(`✗ Video     ${id} is not found`)}`);
      }
    }
  });

cli
  .command('magnet index', 'Index magnet database')
  .alias('index')
  .option('--limit [date]', 'Stop at this date')
  .option('--page [page]', 'Start indexing at this page', {
    construct(t) {
      return t ? +t : 1;
    }
  })
  .action(async (option) => {
    await context.magnetStore.index({
      limit: option.limit ? new Date(option.limit) : undefined,
      startPage: option.page,
      earlyStop: !option.force,
      listener: IndexListener
    });
  });

cli
  .command('magnet list <keyword>', 'Search magnet database')
  .alias('magnet ls')
  .action(async (keyword) => {
    const list = await context.magnetStore.search(keyword);
    printMagnets(list, '');
  });

cli.command('video info <file>', 'Check video info').action(async (file) => {
  const { getVideoInfo } = await import('./video');
  const info = await getVideoInfo(file);
  console.log(JSON.stringify(info, null, 2));
});

cli
  .command('user create', 'Create a new token')
  .option('--comment [comment]', 'Comment of the new token')
  .option('--type [type]', 'One of admin or user')
  .action(async (option) => {
    const { AdminClient } = await import('./client');
    const client = await AdminClient.create();
    const token = await client.createToken({
      comment: option.comment,
      type: option.type === 'admin' ? 'admin' : 'user'
    });
    if (token) {
      console.log(`  ${green(`✓ Create token OK`)}`);
      console.log(`    ${dim('Token')}   ${token.token}`);
      console.log(`    ${dim('Type')}    ${token.type}`);
      console.log(
        `    ${dim('Comment')} ${token.comment ? token.comment : '(Empty)'}`
      );
    } else {
      console.log(`  ${red(`✗ Create token fail`)}`);
    }
  });

cli
  .command('user list', 'List user tokens')
  .alias('user ls')
  .action(async () => {
    const { AdminClient } = await import('./client');
    const client = await AdminClient.create();
    const tokens = await client.listToken();
    if (tokens.length > 0) {
      console.log(`  ${green(`✓ There are ${tokens.length} tokens`)}`);
      for (const token of tokens) {
        const comment =
          token.type === 'visitor' && token.access?.length
            ? dim(`(IP: ${token.access[0].ip})`)
            : !!token.comment
            ? dim(`(Comment: ${token.comment})`)
            : '';
        console.log(
          `  ${dim('•')} ${lightBlue(token.type)} ${token.token} ${comment}`
        );
      }
    }
  });

cli
  .command('user remove [token]', 'Remove user tokens')
  .alias('user rm')
  .option('--visitor', 'Clear all the visitor tokens')
  .action(async (token, option) => {
    const { AdminClient } = await import('./client');
    const client = await AdminClient.create();
    if (option.visitor) {
      const tokens = await client.removeVisitors();
      if (tokens !== undefined) {
        console.log(`  ${green(`✓ Remove ${tokens.length} visitor tokens`)}`);
        if (tokens.length > 0) {
          for (const token of tokens) {
            console.log(`  ${dim('•')} ${token}`);
          }
        }
      } else {
        console.log(`  ${red(`✗ Remove visitor tokens fail`)}`);
      }
    } else if (token) {
      const ok = await client.removeToken(token);
      if (ok) {
        console.log(`  ${green(`✓ Remove ${token}`)}`);
      } else {
        console.log(`  ${red(`✗ Remove ${token} fail`)}`);
      }
    }
  });

cli
  .command('space', 'Open AnimePaste space directory and run script on it')
  .action(async (option) => {
    const cmd = option['--'];
    if (cmd.length > 0) {
      execSync(cmd.join(' '), {
        cwd: context.root,
        env: process.env,
        stdio: ['inherit', 'inherit', 'inherit']
      });
    } else {
      console.log(context.root);
      execSync(`code "${context.root}"`, {
        env: process.env,
        stdio: ['pipe', 'pipe', 'pipe'],
        windowsHide: true
      });
    }
  });

function getVersion(): string {
  const pkg = path.join(__dirname, '../package.json');
  if (existsSync(pkg)) {
    return JSON.parse(readFileSync(pkg, 'utf-8')).version;
  } else {
    return JSON.parse(
      readFileSync(path.join(__dirname, '../../package.json'), 'utf-8')
    ).version;
  }
}

export async function bootstrap() {
  const handle = (error: unknown) => {
    if (error instanceof Error) {
      console.error(lightRed('  Error ') + error.message);
    } else {
      console.error(error);
    }
    debug(error);
  };

  process.setMaxListeners(128);
  process.on('unhandledRejection', (error) => {
    debug(error);
  });

  try {
    cli.on('pre', async (option) => {
      await context.init(option);
    });
    await cli.run(process.argv.slice(2));
    process.exit(0);
  } catch (error: unknown) {
    handle(error);
    process.exit(1);
  }
}

bootstrap();
