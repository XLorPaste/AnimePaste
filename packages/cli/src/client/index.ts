import { context } from '../context';

import { LocalSyncClient } from './local';
import { RemoteSyncClient } from './remote';

export * from './remote';

export * from './client';

export * from './types';

export async function initClient() {
  const config = await context.getSyncConfig();
  if (config.remote && config.remote.baseURL && config.remote.token) {
    return await RemoteSyncClient.init();
  } else {
    return await LocalSyncClient.init();
  }
}
