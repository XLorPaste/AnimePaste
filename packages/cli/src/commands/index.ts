import path from 'node:path';
import { execSync } from 'node:child_process';
import { existsSync, readFileSync, remove } from 'fs-extra';

import Breadc from 'breadc';
import { debug as createDebug } from 'debug';
import { red, link, green, dim, lightBlue } from 'kolorist';

import type { AnimeType } from '../types';

import { context } from '../context';
import { printVideoInfo } from '../io';
import { logger, IndexListener, padRight, DOT } from '../logger';

const debug = createDebug('anime:cli');

export const cli = Breadc('anime', {
  version: getVersion(),
  description: 'Paste your favourite anime online.',
  logger: { debug }
}).option('-f, --force', 'Enable force mode and prefer not using cache');

cli
  .command('watch', 'Watch anime resources update')
  .option('-i, --interval [duration]', 'Damon interval in minutes', {
    construct(t) {
      return t ? +t : 10;
    }
  })
  .option('-o, --once', 'Just do an immediate update')
  .option('--update', 'Only update info')
  .action(async (option) => {
    const { startDaemon } = await import('../daemon');
    await startDaemon(option);
  });

cli
  .command('plan')
  .option('--type [type]', 'One of local or server')
  .action(async (option) => {
    const { Plan } = await import('../daemon');
    const plan = await Plan.create();
    const type = option.type ?? 'local';
    if (type === 'local') {
      plan.printOnair();
    } else if (type === 'server') {
      plan.printOnair();
    }
  });

cli
  .command('search [anime]', 'Search Bangumi resources')
  .option('--type [type]', {
    construct(t) {
      if (t && ['tv', 'web', 'movie', 'ova'].includes(t)) {
        return t as AnimeType;
      } else {
        return 'tv';
      }
    }
  })
  .option('--raw', 'Print raw magnets')
  .option('--index', 'Index magnet database')
  .option('-y, --year [year]')
  .option('-m, --month [month]')
  .option('-p, --plan', 'Output plan.yaml')
  .action(async (anime, option) => {
    const { userSearch } = await import('../anime');
    if (option.index) {
      await context.magnetStore.index({ listener: IndexListener });
    }
    await userSearch(anime, option);
  });

cli
  .command(
    'fetch <id> <title> [...keywords]',
    'Fetch resources using Bangumi ID'
  )
  .option('--raw', 'Print raw magnets')
  .option('--index', 'Index magnet database')
  .option('-p, --plan', 'Output plan.yaml')
  .action(async (id, title, anime, option) => {
    const { daemonSearch } = await import('../anime');
    if (option.index) {
      await context.magnetStore.index({ listener: IndexListener });
    }
    await daemonSearch(id, [title, ...anime], {
      ...option,
      title,
      type: 'tv' as 'tv'
    });
  });

cli
  .command('store list [name]', 'List all uploaded video info')
  .alias('store ls')
  .option('--one-line', 'Only show one line')
  .action(async (name, option) => {
    const { useStore } = await import('../io');
    const store = await useStore('ali')();

    const videos = await store.listLocalVideos();
    videos.sort((a, b) => a.title.localeCompare(b.title));

    const logs: string[] = [];
    const ids: string[] = [];
    for (const info of videos) {
      if (!name || info.title.indexOf(name) !== -1) {
        if (option['one-line']) {
          logs.push(info.videoId);
        } else {
          logs.push(`${info.title}`);
          ids.push(`(${link(info.videoId, info.playUrl[0])})`);
        }
      }
    }

    if (option['one-line']) {
      logger.println(logs.join(' '));
    } else {
      const padded = padRight(logs);
      for (let i = 0; i < padded.length; i++) {
        logger.println(`${DOT} ${padded[i]}  ${ids[i]}`);
      }
    }
  });

cli.command('store info <id>', 'Print video info on OSS').action(async (id) => {
  const { useStore, printVideoInfo } = await import('../io');
  const store = await useStore('ali')();

  const info = await store.fetchVideoInfo(id);

  if (info) {
    printVideoInfo(info);
  } else {
    logger.println(`${red(`✗ video "${id}" not found`)}`);
  }
});

cli
  .command('store put <file>', 'Upload video to OSS')
  .action(async (filename) => {
    const { useStore, printVideoInfo } = await import('../io');
    const store = await useStore('ali')();

    try {
      const info = await store.upload(path.resolve(filename));
      if (info) {
        printVideoInfo(info);
      } else {
        throw new Error();
      }
    } catch (error) {
      logger.empty();
      logger.println(`${red('✗ Fail')} uploading ${filename}`);
    }
  });

cli
  .command('store remove [...ids]', 'Remove video info on OSS')
  .alias('store rm')
  .option('--local', 'Remove local videos')
  .action(async (ids, option) => {
    const { useStore } = await import('../io');
    const store = await useStore('ali')();

    for (const id of ids) {
      const info = await store.fetchVideoInfo(id);

      logger.empty();
      if (info) {
        printVideoInfo(info);
        await store.deleteVideo(info.videoId);
        logger.println(`${green(`✓ Delete    ${info.videoId}`)}`);
      } else {
        logger.println(`${red(`✗ Video     ${id} is not found`)}`);
      }

      if (option.local && info?.source.directory) {
        const filepath = context.decodePath(info?.source.directory, info.title);
        await remove(filepath);
        logger.println(`${green(`✓ Delete    ${filepath}`)}`);
      }
    }
  });

cli
  .command('magnet index', 'Index magnet database')
  .alias('index')
  .option('--limit [date]', 'Stop at this date')
  .option('--page [page]', 'Start indexing at this page', {
    construct(t) {
      return t ? +t : 1;
    }
  })
  .action(async (option) => {
    await context.magnetStore.index({
      limit: option.limit ? new Date(option.limit) : undefined,
      startPage: option.page,
      earlyStop: !option.force,
      listener: IndexListener
    });
  });

cli
  .command('magnet list <keyword>', 'Search magnet database')
  .alias('magnet ls')
  .action(async (keyword) => {
    const magnets = await context.magnetStore.search(keyword);
    magnets.sort((a, b) => a.title.localeCompare(b.title));
    for (const item of magnets) {
      logger.println(
        `${DOT} ${link(item.title, context.magnetStore.idToLink(item.id))}`
      );
    }
  });

cli.command('video info <file>', 'Check video info').action(async (file) => {
  const { getVideoInfo } = await import('../video');
  const info = await getVideoInfo(file);
  console.log(JSON.stringify(info, null, 2));
});

cli
  .command('user create', 'Create a new token')
  .option('--comment [comment]', 'Comment of the new token')
  .option('--type [type]', 'One of admin or user')
  .action(async (option) => {
    const { AdminClient } = await import('../client');
    const client = await AdminClient.create();
    const token = await client.createToken({
      comment: option.comment,
      type: option.type === 'admin' ? 'admin' : 'user'
    });
    if (token) {
      logger.println(`${green(`✓ Create token OK`)}`);
      logger.tab.println(`${dim('Token')}   ${token.token}`);
      logger.tab.println(`${dim('Type')}    ${token.type}`);
      logger.tab.println(
        `${dim('Comment')} ${token.comment ? token.comment : '(Empty)'}`
      );
    } else {
      logger.println(`${red(`✗ Create token fail`)}`);
    }
  });

cli
  .command('user list', 'List user tokens')
  .alias('user ls')
  .action(async () => {
    const { AdminClient } = await import('../client');
    const client = await AdminClient.create();
    const tokens = await client.listToken();
    if (tokens.length > 0) {
      logger.println(`${green(`✓ There are ${tokens.length} tokens`)}`);
      for (const token of tokens) {
        const comment =
          token.type === 'visitor' && token.access?.length
            ? dim(`(IP: ${token.access[0].ip})`)
            : !!token.comment
            ? dim(`(Comment: ${token.comment})`)
            : '';
        logger.println(
          `${DOT} ${lightBlue(token.type)} ${token.token} ${comment}`
        );
      }
    }
  });

cli
  .command('user remove [token]', 'Remove user tokens')
  .alias('user rm')
  .option('--visitor', 'Clear all the visitor tokens')
  .action(async (token, option) => {
    const { AdminClient } = await import('../client');
    const client = await AdminClient.create();
    if (option.visitor) {
      const tokens = await client.removeVisitors();
      if (tokens !== undefined) {
        logger.println(`${green(`✓ Remove ${tokens.length} visitor tokens`)}`);
        if (tokens.length > 0) {
          for (const token of tokens) {
            logger.println(`${DOT} ${token}`);
          }
        }
      } else {
        logger.println(`${red(`✗ Remove visitor tokens fail`)}`);
      }
    } else if (token) {
      const ok = await client.removeToken(token);
      if (ok) {
        logger.println(`${green(`✓ Remove ${token}`)}`);
      } else {
        logger.println(`${red(`✗ Remove ${token} fail`)}`);
      }
    }
  });

cli
  .command('space', 'Open AnimePaste space directory and run script on it')
  .action(async (option) => {
    const cmd = option['--'];
    if (cmd.length > 0) {
      execSync(cmd.join(' '), {
        cwd: context.root,
        env: process.env,
        stdio: ['inherit', 'inherit', 'inherit']
      });
    } else {
      console.log(context.root);
      execSync(`code "${context.root}"`, {
        env: process.env,
        stdio: ['pipe', 'pipe', 'pipe'],
        windowsHide: true
      });
    }
  });

function getVersion(): string {
  const pkg = path.join(__dirname, '../package.json');
  if (existsSync(pkg)) {
    return JSON.parse(readFileSync(pkg, 'utf-8')).version;
  } else {
    return JSON.parse(
      readFileSync(path.join(__dirname, '../../package.json'), 'utf-8')
    ).version;
  }
}
