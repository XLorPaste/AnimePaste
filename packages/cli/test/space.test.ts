import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { describe, it, expect } from 'vitest';

import { makeSystem } from '../src';

const __dirname = path.join(fileURLToPath(import.meta.url), '../');

describe('system', () => {
  it('should load', async () => {
    const root = path.join(__dirname, '../../core/test/fixtures/space');
    const system = await makeSystem(root);
    expect(await system.space.plans()).toEqual([
      {
        date: new Date('2023-04-01T13:00:00.000Z'),
        name: '2023-04-04 新番放送计划',
        onair: [
          {
            bgmId: '323651',
            date: new Date('2023-04-01T13:00:00.000Z'),
            fansub: ['Lilith-Raws', 'ANi'],
            keywords: {
              exclude: [],
              include: [['熊熊勇闯异世界 Punch!']]
            },
            season: 2,
            state: 'onair',
            title: '熊熊勇闯异世界 Punch!',
            translations: {},
            type: '番剧'
          },
          {
            bgmId: '404804',
            date: new Date('2023-04-01T13:00:00.000Z'),
            fansub: ['SweetSub'],
            keywords: {
              exclude: [],
              include: [['天国大魔境', 'Tengoku Daimakyou']]
            },
            season: 1,
            state: 'onair',
            title: '天国大魔境',
            translations: {
              unknown: ['Tengoku Daimakyou']
            },
            type: '番剧'
          },
          {
            bgmId: '376703',
            date: new Date('2023-04-01T13:00:00.000Z'),
            fansub: ['喵萌奶茶屋'],
            keywords: {
              exclude: ['闪耀色彩'],
              include: [['偶像大师', 'iDOLM@STER'], ['灰姑娘女孩'], ['U149']]
            },
            season: 1,
            state: 'onair',
            title: '偶像大师 灰姑娘女孩 U149',
            translations: {},
            type: '番剧'
          }
        ],
        state: 'onair'
      }
    ]);
  });
});
