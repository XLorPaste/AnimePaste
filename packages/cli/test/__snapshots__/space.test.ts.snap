// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`system > should load 1`] = `
[
  {
    "date": 2023-04-01T13:00:00.000Z,
    "name": "2023-04-04 新番放送计划",
    "onair": [
      {
        "alias": [],
        "bgm": "323651",
        "date": 2023-04-01T13:00:00.000Z,
        "fansub": [
          "Lilith-Raws",
          "ANi",
        ],
        "keywords": {
          "exclude": [],
          "include": [
            [
              "熊熊勇闯异世界 Punch!",
            ],
          ],
        },
        "preference": {
          "extension": {
            "exclude": [],
            "include": [
              "mp4",
              "mkv",
            ],
          },
          "format": {
            "episode": "[{fansub}] {title} - S{season}E{ep}.{extension}",
            "film": "[{fansub}] {title}.{extension}",
            "ova": "[{fansub}] {title}.{extension}",
          },
          "keyword": {
            "exclude": [],
            "order": {
              "format": [
                "mp4",
                "mkv",
              ],
              "language": [
                "简",
                "繁",
              ],
              "resolution": [
                "1080",
                "720",
              ],
            },
          },
        },
        "season": 2,
        "status": "onair",
        "storage": {
          "name": "anime",
          "root": Path { fs: "node", path: "../core/test/fixtures/space/anime" },
        },
        "title": "熊熊勇闯异世界 Punch!",
        "translations": {},
        "type": "番剧",
      },
      {
        "alias": [],
        "bgm": "404804",
        "date": 2023-04-01T13:00:00.000Z,
        "fansub": [
          "SweetSub",
        ],
        "keywords": {
          "exclude": [],
          "include": [
            [
              "天国大魔境",
              "Tengoku Daimakyou",
            ],
          ],
        },
        "preference": {
          "extension": {
            "exclude": [],
            "include": [
              "mp4",
              "mkv",
            ],
          },
          "format": {
            "episode": "[{fansub}] {title} - S{season}E{ep}.{extension}",
            "film": "[{fansub}] {title}.{extension}",
            "ova": "[{fansub}] {title}.{extension}",
          },
          "keyword": {
            "exclude": [],
            "order": {
              "format": [
                "mp4",
                "mkv",
              ],
              "language": [
                "简",
                "繁",
              ],
              "resolution": [
                "1080",
                "720",
              ],
            },
          },
        },
        "season": 1,
        "status": "onair",
        "storage": {
          "name": "anime",
          "root": Path { fs: "node", path: "../core/test/fixtures/space/anime" },
        },
        "title": "天国大魔境",
        "translations": {
          "unknown": [
            "Tengoku Daimakyou",
          ],
        },
        "type": "番剧",
      },
      {
        "alias": [],
        "bgm": "376703",
        "date": 2023-04-01T13:00:00.000Z,
        "fansub": [
          "喵萌奶茶屋",
        ],
        "keywords": {
          "exclude": [
            "闪耀色彩",
          ],
          "include": [
            [
              "偶像大师 灰姑娘女孩 U149",
            ],
            [
              "偶像大师",
              "iDOLM@STER",
            ],
            [
              "灰姑娘女孩",
            ],
            [
              "U149",
            ],
          ],
        },
        "preference": {
          "extension": {
            "exclude": [],
            "include": [
              "mp4",
              "mkv",
            ],
          },
          "format": {
            "episode": "[{fansub}] {title} - S{season}E{ep}.{extension}",
            "film": "[{fansub}] {title}.{extension}",
            "ova": "[{fansub}] {title}.{extension}",
          },
          "keyword": {
            "exclude": [],
            "order": {
              "format": [
                "mp4",
                "mkv",
              ],
              "language": [
                "简",
                "繁",
              ],
              "resolution": [
                "1080",
                "720",
              ],
            },
          },
        },
        "season": 1,
        "status": "onair",
        "storage": {
          "name": "anime",
          "root": Path { fs: "node", path: "../core/test/fixtures/space/anime" },
        },
        "title": "偶像大师 灰姑娘女孩 U149",
        "translations": {},
        "type": "番剧",
      },
    ],
    "preference": {
      "extension": {
        "exclude": [],
        "include": [
          "mp4",
          "mkv",
        ],
      },
      "format": {
        "episode": "[{fansub}] {title} - S{season}E{ep}.{extension}",
        "film": "[{fansub}] {title}.{extension}",
        "ova": "[{fansub}] {title}.{extension}",
      },
      "keyword": {
        "exclude": [],
        "order": {
          "format": [
            "mp4",
            "mkv",
          ],
          "language": [
            "简",
            "繁",
          ],
          "resolution": [
            "1080",
            "720",
          ],
        },
      },
    },
    "status": "onair",
  },
]
`;
